File: go/internal/agent/fixflow/registry_test.go
package fixflow

import (
	"sync"
	"sync/atomic"
	"testing"
	"time"
)

func noTimeout(string) {}

func TestRegistryResolveOnce(t *testing.T) {
	r := newRunRegistry()
	r.Park("o/r#1", &ParkedRun{SessionID: "s", CallID: "c"}, time.Hour, noTimeout)
	if r.Len() != 1 {
		t.Fatalf("len = %d", r.Len())
	}

	run, ok := r.Resolve("o/r#1")
	if !ok || run.CallID != "c" {
		t.Fatalf("first resolve = %+v, %v", run, ok)
	}
	if _, ok := r.Resolve("o/r#1"); ok {
		t.Error("second resolve should find nothing (already claimed)")
	}
	if r.Len() != 0 {
		t.Errorf("len after resolve = %d", r.Len())
	}
}

func TestRegistryLateResolveNoop(t *testing.T) {
	r := newRunRegistry()
	if _, ok := r.Resolve("never/parked#9"); ok {
		t.Error("resolving an unparked PR should no-op")
	}
}

// TestRegistryTimeoutClaims proves the timer path claims the run, and a webhook that
// arrives afterward (late) finds nothing.
func TestRegistryTimeoutClaims(t *testing.T) {
	r := newRunRegistry()
	claimed := make(chan bool, 1)
	r.Park("o/r#2", &ParkedRun{SessionID: "s", CallID: "c"}, 10*time.Millisecond, func(pr string) {
		_, ok := r.Resolve(pr)
		claimed <- ok
	})

	select {
	case ok := <-claimed:
		if !ok {
			t.Fatal("timeout fired but failed to claim the parked run")
		}
	case <-time.After(time.Second):
		t.Fatal("timeout never fired")
	}
	// Late webhook after the timeout claimed it.
	if _, ok := r.Resolve("o/r#2"); ok {
		t.Error("late webhook after timeout should find nothing")
	}
}

// TestRegistryConcurrentResolveExactlyOne proves the atomic claim under contention:
// when many callers (duplicate webhooks + a timer) race to resolve one parked run,
// exactly one wins.
func TestRegistryConcurrentResolveExactlyOne(t *testing.T) {
	r := newRunRegistry()
	r.Park("o/r#3", &ParkedRun{SessionID: "s", CallID: "c"}, time.Hour, noTimeout)

	const racers = 50
	var wins int64
	var wg sync.WaitGroup
	start := make(chan struct{})
	for i := 0; i < racers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			<-start
			if _, ok := r.Resolve("o/r#3"); ok {
				atomic.AddInt64(&wins, 1)
			}
		}()
	}
	close(start)
	wg.Wait()

	if wins != 1 {
		t.Fatalf("exactly one resolver must win, got %d", wins)
	}
}

// TestRegistryRepark proves a retry re-parks under the same PR key (stopping the old
// timer), and only the latest parking is resolvable.
func TestRegistryRepark(t *testing.T) {
	r := newRunRegistry()
	r.Park("o/r#4", &ParkedRun{SessionID: "s", CallID: "c1", Attempts: 1}, time.Hour, noTimeout)
	r.Park("o/r#4", &ParkedRun{SessionID: "s", CallID: "c2", Attempts: 2}, time.Hour, noTimeout)
	if r.Len() != 1 {
		t.Fatalf("re-park should replace, len = %d", r.Len())
	}
	run, ok := r.Resolve("o/r#4")
	if !ok || run.CallID != "c2" || run.Attempts != 2 {
		t.Fatalf("resolve = %+v, %v (want latest c2/2)", run, ok)
	}
}

File: kotlin/src/test/kotlin/io/github/jkjamies/automationagent/agent/fixflow/RegistryTest.kt
package io.github.jkjamies.automationagent.agent.fixflow

import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeout
import java.util.concurrent.atomic.AtomicInteger
import kotlin.time.Duration.Companion.hours
import kotlin.time.Duration.Companion.milliseconds

private fun newRegistry(): RunRegistry = RunRegistry(CoroutineScope(SupervisorJob() + Dispatchers.Default))

private val noTimeout: suspend (String) -> Unit = { }

class RegistryTest : BehaviorSpec({
    Given("a parked run") {
        When("resolving it twice") {
            Then("the first wins and the second finds nothing") {
                val r = newRegistry()
                r.park("o/r#1", ParkedRun("s", "c"), 1.hours, noTimeout)
                r.size() shouldBe 1
                r.resolve("o/r#1").shouldNotBeNull().callId shouldBe "c"
                r.resolve("o/r#1") shouldBe null
                r.size() shouldBe 0
            }
        }
    }

    Given("an unparked PR") {
        When("resolving it") {
            Then("it is a no-op") {
                newRegistry().resolve("never/parked#9") shouldBe null
            }
        }
    }

    Given("a parked run whose timer fires") {
        When("the timeout claims it") {
            Then("a late webhook afterward finds nothing") {
                val r = newRegistry()
                val claimed = CompletableDeferred<Boolean>()
                r.park("o/r#2", ParkedRun("s", "c"), 10.milliseconds) { pr -> claimed.complete(r.resolve(pr) != null) }
                withTimeout(2000) { claimed.await() } shouldBe true
                r.resolve("o/r#2") shouldBe null // late webhook after the timeout claimed it
            }
        }
    }

    Given("many racers resolving one parked run") {
        When("they contend") {
            Then("exactly one wins") {
                val r = newRegistry()
                r.park("o/r#3", ParkedRun("s", "c"), 1.hours, noTimeout)
                val wins = AtomicInteger(0)
                coroutineScope {
                    repeat(50) { launch { if (r.resolve("o/r#3") != null) wins.incrementAndGet() } }
                }
                wins.get() shouldBe 1
            }
        }
    }

    Given("a re-park under the same PR key") {
        When("resolving") {
            Then("only the latest parking is resolvable") {
                val r = newRegistry()
                r.park("o/r#4", ParkedRun("s", "c1", attempts = 1), 1.hours, noTimeout)
                r.park("o/r#4", ParkedRun("s", "c2", attempts = 2), 1.hours, noTimeout)
                r.size() shouldBe 1
                val run = r.resolve("o/r#4").shouldNotBeNull()
                run.callId shouldBe "c2"
                run.attempts shouldBe 2
            }
        }
    }
})
